lib/msal-common/src/utils/ProtocolUtils.ts

```diff
@@ -15,23 +28,56 @@ export class ProtocolUtils {
      * @param userState 
      * @param randomGuid 
      */
-    static setRequestState(userState: string, randomGuid: string): string {
-        return !StringUtils.isEmpty(userState) ? `${randomGuid}${Constants.RESOURCE_DELIM}${userState}` : randomGuid;
+    static setRequestState(userState: string, randomGuid: string, cryptoObj: ICrypto): string {
+        const libraryState = ProtocolUtils.generateLibraryState(randomGuid, cryptoObj);
+        return !StringUtils.isEmpty(userState) ? `${libraryState}${Constants.RESOURCE_DELIM}${userState}` : libraryState;
     }
 
     /**
-     *
-     * Extracts user state value from the state sent with the authentication request.
-     * @returns {string} scope.
-     * @ignore
+     * Generates the state value used by the library.
+     * @param randomGuid 
+     * @param cryptoObj 
+     * 
+     * @returns Base64 encoded string representing the state
      */
-    static getUserRequestState(serverResponseState: string): string {
-        if (!StringUtils.isEmpty(serverResponseState)) {
-            const splitIndex = serverResponseState.indexOf(Constants.RESOURCE_DELIM);
-            if (splitIndex > -1 && splitIndex + 1 < serverResponseState.length) {
-                return serverResponseState.substring(splitIndex + 1);
+    static generateLibraryState(randomGuid: string, cryptoObj: ICrypto): string {
+        const stateObject: LibraryStateObject = {
+            id: randomGuid,
+            ts: TimeUtils.nowSeconds()
+        };
+
+        const stateString = JSON.stringify(stateObject);
+
+        return cryptoObj.base64Encode(stateString);
+    }
+
+    /**
+     * Extracts user state value and library state value 
+     * from the state sent with the authentication request.
+     * @param state 
+     * @param cryptoObj 
+     */
+    static parseRequestState(state: string, cryptoObj: ICrypto): RequestStateObject {
+        try {
+            if (StringUtils.isEmpty(state)) {
+                state = "";
             }
+
+            const [
+                libraryState,
+                userState
+            ] = decodeURIComponent(state).split(Constants.RESOURCE_DELIM);
```

Comment:
We should make sure that we are properly handling if `userState` contains the delimiter.

Reply:
I've updated the code to handle this scenario, and added tests for the use case as well.